src/cpprealm/internal/network/network_transport.cpp

```diff
@@ -78,33 +82,99 @@ namespace realm::internal {
+        Context m_ssl_context;
+        DefaultSocket socket{service};
 
         try {
+            realm::sync::network::Endpoint ep;
             auto resolver = realm::sync::network::Resolver{service};
-            auto resolved = resolver.resolve(sync::network::Resolver::Query(host, "443"));
-            ep = *resolved.begin();
+            if (m_proxy_config) {
+                std::string proxy_address = host_from_url(m_proxy_config->address);
+                if (proxy_address.empty()) {
+                    std::error_code e;
+                    auto address = realm::sync::network::make_address(m_proxy_config->address, e);
+                    if (e.value() > 0) {
+                        app::Response response;
+                        response.custom_status_code = e.value();
+                        response.body = e.message();
+                        completion_block(std::move(response));
+                        return;
+                    }
+                    ep = realm::sync::network::Endpoint(address, m_proxy_config->port);
+                } else {
+                    auto resolved = resolver.resolve(sync::network::Resolver::Query(proxy_address, m_proxy_config->port));
+                    ep = *resolved.begin();
+                }
+            } else {
+                auto resolved = resolver.resolve(sync::network::Resolver::Query(host, "443"));
+                ep = *resolved.begin();
+            }
+            socket.connect(ep);
         } catch (...) {
             app::Response response;
-            response.http_status_code = 500;
+            response.custom_status_code = util::error::operation_aborted;
             completion_block(std::move(response));
             return;
         }
 
-        using namespace realm::sync::network::ssl;
-        Context m_ssl_context;
-        DefaultSocket socket{service};
-        socket.connect(ep);
+        auto logger = util::Logger::get_default_logger();
+
+        if (m_proxy_config) {
+            realm::sync::HTTPRequest req;
+            req.method = realm::sync::HTTPMethod::Connect;
+            req.headers.emplace("Host", util::format("%1:%2", host, "443"));
+            if (m_proxy_config->username_password) {
+                auto userpass = util::format("%1:%2", m_proxy_config->username_password->first, m_proxy_config->username_password->second);
+                std::string encoded_userpass;
+                encoded_userpass.resize(realm::util::base64_encoded_size(userpass.length()));
+                realm::util::base64_encode(userpass.data(), userpass.size(), encoded_userpass.data(), encoded_userpass.size());
+                req.headers.emplace("Proxy-Authorization", util::format("Basic %1", encoded_userpass));
+            }
+            if (m_custom_http_headers) {
+                for (auto& header : *m_custom_http_headers) {
+                    req.headers.emplace(header);
+                }
+            }
```

Comment:
We shouldn't be attaching the custom headers to the `CONNECT` request we send the proxy, only the `Proxy-Authorization` header ever makes sense here.